flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java

```diff
@@ -0,0 +1,347 @@
+    }
+
+    @Override
+    public void close() throws IOException {
+        streamProxy.close();
+    }
+
+    private List<KinesisShardSplit> initialDiscoverSplits() {
+        LOG.info("initialDiscoverSplits");
+        List<Shard> shards = streamProxy.listShards(streamArn, lastSeenShardId);
+        return mapToSplits(
+                shards,
+                InitialPosition.valueOf(
+                        consumerConfig
+                                .getOrDefault(
+                                        STREAM_INITIAL_POSITION, DEFAULT_STREAM_INITIAL_POSITION)
+                                .toString()));
+    }
+
+    /**
+     * This method is used to discover Kinesis splits the job can subscribe to. It can be run in
+     * parallel, is important to not mutate any shared state.
+     *
+     * @return list of discovered splits
+     */
+    private List<KinesisShardSplit> periodicallyDiscoverSplits() {
+        List<Shard> shards = streamProxy.listShards(streamArn, lastSeenShardId);
+        // Any shard discovered after the initial startup should be read from the start, since they
+        // come from resharding
+        return mapToSplits(shards, InitialPosition.TRIM_HORIZON);
+    }
+
+    private List<KinesisShardSplit> mapToSplits(
+            List<Shard> shards, InitialPosition initialPosition) {
+        StartingPosition startingPosition;
+        switch (initialPosition) {
+            case LATEST:
+                // If LATEST is requested, we still set the starting position to the time of
+                // startup. This way, the job starts reading from a deterministic timestamp
+                // (i.e. time of job submission), even if it enters a restart loop immediately
+                // after submission.
+                startingPosition = StartingPosition.fromTimestamp(Instant.now());
+                break;
+            case AT_TIMESTAMP:
+                startingPosition =
+                        StartingPosition.fromTimestamp(
+                                parseStreamTimestampStartingPosition(consumerConfig).toInstant());
+                break;
+            case TRIM_HORIZON:
+            default:
+                startingPosition = StartingPosition.fromStart();
+        }
+
+        List<KinesisShardSplit> splits = new ArrayList<>();
+        for (Shard shard : shards) {
+            splits.add(new KinesisShardSplit(streamArn, shard.shardId(), startingPosition));
+        }
+
+        return splits;
+    }
+
+    /**
+     * This method assigns a given set of Kinesis splits to the readers currently registered on the
+     * cluster. This assignment is done via a side-effect on the {@link SplitEnumeratorContext}
+     * object.
+     *
+     * @param discoveredSplits list of discovered splits
+     * @param t throwable thrown when discovering splits. Will be null if no throwable thrown.
+     */
+    private void assignSplits(List<KinesisShardSplit> discoveredSplits, Throwable t) {
```

Comment:
nit: Instead of `t` use a meaningful arg name

Reply:
done.